docs/metasploit-framework.wiki/Java-Meterpreter-Feature-Parity-Proposal.md

```diff
@@ -0,0 +1,253 @@
+  disk - discussed further below
+- Update CI/Maven build steps
+
+Alternative implementation steps are also documented.
+
+## Implementation
+
+### Supporting native system calls
+
+We will move the OS detection from `stdapi` to `core`. This would allow us to detect the victim's OS and architecture as
+part of the `core` API, allowing stdapi to additionally include the correct Railgun and the JNA dll/so/dylib library
+files.
+
+Native call support will be provided by JNA - an open source library which provides easy access to Window's APIs. Using
+JNA would help reduce the boilerplate for making Windows API calls, is maintained by an existing community, and is less
+likely to be detected as malicious.
+
+The sequence of steps required for loading stdapi and invoking clear event log:
+
+```mermaid
+sequenceDiagram
+    msfconsole->>+meterpreter: load core library
+    meterpreter-->>-msfconsole: return success and list of available commands
+
+    msfconsole->>+meterpreter: get architecture from core library
+    meterpreter-->>-msfconsole: e.g. Windows 10 x64
+
+    msfconsole->>+meterpreter: load stdapi - i.e. classfiles + JNA + Railgun dll
+    meterpreter->>meterpreter: Load new java commands
+    note right of meterpreter: Keep JNA + Railgun library in memory<br />Don't load them yet
+    meterpreter-->>-msfconsole: return success and list of available commands
+
+    msfconsole->>+meterpreter: clear event log
+
+    rect rgb(191, 223, 255, .3)
+    note right of meterpreter: Load JNA if it's <br >not been loaded before
+    meterpreter->>meterpreter: Copy JNA from classpath to file system
+    meterpreter->>meterpreter: System.load(temp path)
+    meterpreter->>meterpreter: delete temp path
+    end
+
+    meterpreter->>meterpreter: clear event log using JNA
+    meterpreter-->>-msfconsole: clear event log result
+```
+
+### Railgun support
+
+Railgun requires access to low-level functionality, i.e. directly manipulating memory etc. Java does not support this
+functionality directly unless a Java wrapper is provided. Therefore a JNI wrapper for the current Railgun implementation
+will be developed - similar to the previous [prototype](https://github.com/rapid7/metasploit-payloads/pull/529). Maven
+would be updated to build this library for different architectures/platforms.
+
+Similar to the implementation of native system calls; We will move the OS detection from `stdapi` to `core`. This would
+allow us to detect the victim's OS and architecture as part of the `core` API, allowing `stdapi` to additionally include
+the correct Railgun and JNA dll/so/dylib library files. Once a native call needs to be executed, Meterpreter would
+attempt to load JNA and use it to the native Windows API to begin the process of reflectively loading the compiled
+Railgun library.
+
+The sequence of steps required for loading stdapi and invoking Railgun:
+
+```mermaid
```

Comment:
This doesn't render for now, but I'll not treat it as a blocker for now - and I'll circle back to getting mermaid diagrams wired up :+1: